### ISSUE TITLE
Trying to set the request-id header on a HttpClient request results in unexpected header

### ISSUE THREAD

**Issue by karlra** [dotnet/runtime#35357]
@davidfowl asked me to create a new issue in this repository mirroring this: https://github.com/dotnet/aspnetcore/issues/21012

### Describe the bug

If, in an ASP.net core project, you create an instance of HttpClient and make a request with a manually set header called "request-id", the header that arrives at the server isn't what you expect.

### How to reproduce

1. Create an ASP.net core project
2. In a controller method (or anywhere after ASP.net core has been initialized), run the following:

```c#
HttpClient client = new HttpClient();
StringContent content = new StringContent("hi!");
content.Headers.Add("request-id", "hello!");
await client.PostAsync("http://localhost:5555", content);
```

It doesn't happen if you make the request before asp.net has been initialized.

### Expected result

When the request arrives at the server, it includes a header called "request-id" with the value **hello!**

### Actual result

When the request arrives at the server, it includes a header called "request-id" with the value **|8da1737a-4c14b253bd8e4130.1., hello!**

The |8da1737a-4c14b253bd8e4130.1 part is dynamic and changes from request to request.

### Full repro

1. Create a new ASP.net Core project 
2. Delete all code files
3. Paste in the following for Program.cs 
4. Change the "localhost:5555" part to somewhere you can see the requests arrive
5. Run program

```c#
using System.Net.Http;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Hosting;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

namespace WebApplicationRequestIdBug
{
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();   
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }

    public class Startup
    {
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            app.Run(async context =>
            {
                HttpClient client = new HttpClient();
                StringContent content = new StringContent("hi!");
                content.Headers.Add("request-id", "hello!");
                await client.PostAsync("http://localhost:5555", content);
            });
        }
    }
}

```

**Comment by davidfowl** [comment]
> In a controller method (or anywhere after ASP.net core has been initialized), run the following:

That's inaccurate, it needs to be within the request itself.

A minimal repro without ASP.NET Core should look like this:

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

namespace ConsoleApp11
{
    class Program
    {
        static async Task Main(string[] args)
        {
            var listener = new HttpListener();
            listener.Prefixes.Add("http://localhost:5555/");
            listener.Start();

            async Task Run()
            {
                while (true)
                {
                    var context = await listener.GetContextAsync();

                    Console.WriteLine(context.Request.Headers["Request-Id"]);
                }
            }

            var runningTask = Run();

            var activity = new Activity("Outgoing");
            activity.Start();
            HttpClient client = new HttpClient();
            StringContent content = new StringContent("hi!");
            content.Headers.Add("request-id", "hello!");
            await client.PostAsync("http://localhost:5555", content);
            activity.Stop();

            await runningTask;
        }
    }
}
```

That should apply in a console application as well.

**Comment by davidfowl** [comment]
cc @lmolkova @SergeyKanzhelev

**Comment by karelz** [comment]
Triage:
- As a workaround, you can use `SocketsHttpHandler` directly - that will avoid any `DiagnosticHandler` headers
- Longer-term, there could be a solution on `HttpClientHandler` which is tracked in #35337
- We are ok to change `DiagnosticHandler` implementation to not set its value when it has been set by the user

**Comment by MihaZupan** [comment]
This has been fixed by dotnet/corefx#34321 for Core 3 and backported to 2.1 in dotnet/corefx#34300.
